src/station.cpp

```diff
@@ -305,79 +342,106 @@ Rect Station::GetCatchmentRect() const
 	return ret;
 }
 
-/** Rect and pointer to IndustryVector */
-struct RectAndIndustryVector {
-	Rect rect;                       ///< The rectangle to search the industries in.
-	IndustryVector *industries_near; ///< The nearby industries.
-};
-
 /**
- * Callback function for Station::RecomputeIndustriesNear()
- * Tests whether tile is an industry and possibly adds
- * the industry to station's industries_near list.
- * @param ind_tile tile to check
- * @param user_data pointer to RectAndIndustryVector
- * @return always false, we want to search all tiles
+ * Add nearby industry to station's industries_near list if it accepts cargo.
+ * @param ind Industry
+ * @param st Station
  */
-static bool FindIndustryToDeliver(TileIndex ind_tile, void *user_data)
+static void AddIndustryToDeliver(Industry *ind, Station *st)
 {
-	/* Only process industry tiles */
-	if (!IsTileType(ind_tile, MP_INDUSTRY)) return false;
-
-	RectAndIndustryVector *riv = (RectAndIndustryVector *)user_data;
-	Industry *ind = Industry::GetByTile(ind_tile);
-
 	/* Don't check further if this industry is already in the list */
-	if (riv->industries_near->Contains(ind)) return false;
-
-	/* Only process tiles in the station acceptance rectangle */
-	int x = TileX(ind_tile);
-	int y = TileY(ind_tile);
-	if (x < riv->rect.left || x > riv->rect.right || y < riv->rect.top || y > riv->rect.bottom) return false;
+	if (st->industries_near.Contains(ind)) return;
 
 	/* Include only industries that can accept cargo */
 	uint cargo_index;
 	for (cargo_index = 0; cargo_index < lengthof(ind->accepts_cargo); cargo_index++) {
 		if (ind->accepts_cargo[cargo_index] != CT_INVALID) break;
 	}
-	if (cargo_index >= lengthof(ind->accepts_cargo)) return false;
+	if (cargo_index >= lengthof(ind->accepts_cargo)) return;
 
-	*riv->industries_near->Append() = ind;
+	st->industries_near.Include(ind);
+}
 
+/**
+ * Remove this station from the nearby stations lists of all towns and industries.
+ */
+void Station::RemoveFromAllNearbyLists()
+{
+	Town *t;
+	FOR_ALL_TOWNS(t) t->stations_near.Erase(this);
+	Industry *i;
+	FOR_ALL_INDUSTRIES(i) i->stations_near.Erase(this);
```

Comment:
macro for loops without braces make me uncomfortable

Reply:
Ok, but we do have existing code that does this.